### PR TITLE
[MRG] Added `BatchSimulate` class for batch simulations

### LINKED ISSUE #140: add functionality for running batch simulations
Enable users to run a non-interactive batch of simulations of the HNN model by specifying a set of parameters values to use. This could be used for sweeping a range of parameter values (e.g. optimization), a list of specific parameter files to iterate over, and/or a distribution to draw from in choosing parameter values (e.g. priors for SNPE).

This will be a new interface for running simulations via `hnn_core.simulator` with the parameters to use specified as input and the results retrievable by their respective Python objects, and optionally saving the results/parameter files to disk.

Investigate what can be learned from how it is done in `scipy` and [NetPyNE batch simulations](http://www.netpyne.org/tutorial.html#tutorial-8-running-batch-simulations).

### PR DESCRIPTION
**Pull request by samadpls**
Closes #140

### CODE REVIEW

File: hnn_core/batchsimulate.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            The number of trials for the simulation. Default is 1.
+        """
+        self.set_params = set_params
+        self.net_name = net_name
+        self.tstop = tstop
+        self.dt = dt
+        self.n_trials = n_trials
+
+    def run(self, param_grid, max_size=None, return_output=True,
+            save_output=False, fpath='./', n_jobs=-1):
```
Comment (jasmainak) [inline hnn_core/batchsimulate.py:35]: n_jobs = -1 is dangerous ... you'll choke your computer. Default should always be 1

**Comment by jasmainak** [comment]
I meant that you should add it to the examples gallery :)

**Comment by jasmainak** [comment]
As it stands ... the BatchSimulate does not really enable user to do more things by writing less code ... you need to brainstorm a bit how to make it more useful. Write the API first before starting to code.

```py
batch = BatchSimulate(func)
dpls = batch.simulate()
batch.plot_dpl()
batch.plot_psd()
```

I'm just dreaming ... but think of ways people may want to use such an API and how to simplify boilerplate code for them

**Comment by ntolley** [comment]
@jasmainak totally agree, I think the main utiluty of this function will be organizing code for saving/loading the potentially very large output files, along with some simulation metadata

To make this use less code, I think the best way will be to decide on a limited set of predefined functions. Perhaps even a class that helps users quickly specify the parameters they care about:
```py
net = jones_2009_model()
func = make_evoked_simulator(net, weight_params='ampa', time_params='mean', custom_remove=['evprox1_L2Pyr_ampa'])
batch = BatchSimulate(func)
dpls = batch.simulate()
```
However, I want to avoid this `make_simulator()` scheme being too broad as it could quickly blow up in complexity and no be useful. The safest route right now will be to hard code the parameters for one of these functions, and have the example import the function.

Perhaps this will be a good way to replace this function too!
https://github.com/jonescompneurolab/hnn-core/blob/aa773ec4ca89d92e700883d6756fa3814933d31f/hnn_core/network_models.py#L312

File: examples/howto/batch_simulate.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import numpy as np
+from hnn_core import BatchSimulate
+from hnn_core.network_models import jones_2009_model
+
+# The number of cores may need modifying depending on your current machine.
+n_jobs = 10
+###############################################################################
+
+
+def set_params(param_grid, net=None):
```
Comment (ntolley) [inline examples/howto/batch_simulate.py:29]:
```suggestion
def set_params(param_values, net=None):
```

File: examples/howto/batch_simulate.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                         synaptic_delays=synaptic_delays)
+
+###############################################################################
+
+
+param_grid = {
+    'weight_basket': np.linspace(0.1, 1.0, 5),
+    'weight_pyr': np.linspace(0.5, 2.0, 5),
+    'mu': np.linspace(1, 6, 5),
+    'sigma': np.linspace(1, 3, 5)
```
Comment (ntolley) [inline examples/howto/batch_simulate.py:70]:
```suggestion
    'weight_basket': np.logspace(-4 -1, 5),
    'weight_pyr': np.logspace(-4, -1, 5),
    'mu': np.linspace(20, 80, 5),
    'sigma': np.linspace(1, 20, 5)
```

File: hnn_core/batchsimulate.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        n_trials : int, optional
+            The number of trials for the simulation. Default is 1.
+        """
+        self.set_params = set_params
+        self.net_name = net_name
+        self.tstop = tstop
+        self.dt = dt
+        self.n_trials = n_trials
+
+    def run(self, param_grid, max_size=None, return_output=True,
```
Comment (ntolley) [inline hnn_core/batchsimulate.py:34]:
```suggestion
    def run(self, param_values, max_size=None, return_output=True,
```

File: hnn_core/batchsimulate.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        save_output : bool, optional
+            Whether to save the outputs to disk. Default is False.
+        fpath : str, optional
+            File path for saving outputs. Default is './'.
+        n_jobs : int, optional
+            Number of parallel jobs. Default is -1.
+
+        Returns
+        -------
+        list
```
Comment (ntolley) [inline hnn_core/batchsimulate.py:56]:
```suggestion
        results : list
```

Comment (ntolley) [reply hnn_core/batchsimulate.py:56]:
```py
[{'net': net_sim1, 'dpl': dpl_sim1},
 {'net': net_sim2 'dpl': dpl_sim2]
```

File: hnn_core/batchsimulate.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            net = jones_2009_model()
+        elif self.net_name == 'law':
+            net = law_2021_model()
+        elif self.net_name == 'calcium':
+            net = calcium_model()
+
+        self.set_params(param_values, net)
+        dpl = simulate_dipole(net, tstop=self.tstop, dt=self.dt,
+                              n_trials=self.n_trials)
+        return dpl
```
Comment (ntolley) [inline hnn_core/batchsimulate.py:121]:
```suggestion
        return {'net': net, 'dpl': dpl}
```

File: hnn_core/batchsimulate.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            The number of trials for the simulation. Default is 1.
+        """
+        self.set_params = set_params
+        self.net_name = net_name
+        self.tstop = tstop
+        self.dt = dt
+        self.n_trials = n_trials
+
+    def run(self, param_grid, max_size=None, return_output=True,
+            save_output=False, fpath='./', n_jobs=1):
```
Comment (ntolley) [inline hnn_core/batchsimulate.py:35]:
```suggestion
            save_output=False, combinations=True, fpath='./', n_jobs=1):
```

File: hnn_core/batchsimulate.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        n_jobs : int, optional
+            Number of parallel jobs. Default is -1.
+
+        Returns
+        -------
+        list
+            List of simulation results if return_output is True.
+        """
+        param_combinations = self._generate_param_combinations(
+            param_grid, max_size)
```
Comment (ntolley) [inline hnn_core/batchsimulate.py:60]:
```suggestion
            param_grid, max_size, combinations)
```

**Comment by ntolley** [comment]
after discussion with @jasmainak, useful features will include

- [ ] A callable that can be executed after each simulation, and calculates some sort of summary statistic
- [ ] Plotting functions for uni-, bi-, and tri-variate parameter sweeps (including colormaps and subplot grids)

File: examples/howto/batch_simulate.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+"""
+=================================================
+06. Batch Simulation
+=================================================
```
Comment (jasmainak) [inline examples/howto/batch_simulate.py:4]: == should not exceed text

File: hnn_core/batch_simulate.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        return_output : bool, optional
+            Whether to return the simulation outputs. Default is True.
+        save_output : bool, optional
+            Whether to save the outputs to disk. Default is False.
+        combinations : bool, optional
+            Whether to generate the Cartesian product of the parameter ranges.
+            If False, generate combinations based on corresponding indices.
+            Default is True.
+        fpath : str, optional
+            File path for saving outputs. Default is './'.
```
Comment (jasmainak) [inline hnn_core/batch_simulate.py:60]: unused as of now ...

File: hnn_core/tests/test_batch_simulate.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    results = batch_simulate_instance.simulate_batch(param_combinations,
+                                                     n_jobs=2)
+    assert len(results) == len(param_combinations)
+    for result in results:
+        assert 'net' in result
+        assert 'dpl' in result
+        assert 'param_values' in result
+
+
+def test_run(batch_simulate_instance, param_grid):
```
Comment (jasmainak) [inline hnn_core/tests/test_batch_simulate.py:85]: docstring missing

File: hnn_core/tests/test_batch_simulate.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+                             numspikes=1,
+                             location='proximal',
+                             weights_ampa=weights_ampa,
+                             synaptic_delays=synaptic_delays)
+
+    return BatchSimulate(set_params=set_params, tstop=10.)
+
+
+@pytest.fixture
+def param_grid():
```
Comment (jasmainak) [inline hnn_core/tests/test_batch_simulate.py:36]: public function missing docstring

File: hnn_core/batch_simulate.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                 dt=0.025, n_trials=1):
+        """
+        Initialize the BatchSimulate class.
+
+        Parameters
+        ----------
+        set_params : func
+            User-defined function that sets parameters in network drives.
+            `set_params(net, params) -> None`
+        net_name : str, optional
```
Comment (jasmainak) [inline hnn_core/batch_simulate.py:25]:
```suggestion
        net_name : str
```

let's be consistent with rest of repo ... the function signature tells you if it's optional or not

File: hnn_core/tests/test_batch_simulate.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
+# Authors: Abdul Samad Siddiqui <abdulsamadsid1@gmail.com>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+import pytest
+import numpy as np
+from hnn_core import BatchSimulate
```
Comment (jasmainak) [inline hnn_core/tests/test_batch_simulate.py:8]:
```suggestion
import numpy as np

from hnn_core import BatchSimulate
```

File: hnn_core/batch_simulate.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+from joblib import Parallel, delayed
+from hnn_core import simulate_dipole
+from hnn_core.network_models import (jones_2009_model,
+                                     calcium_model, law_2021_model)
+
+
+class BatchSimulate:
+    def __init__(self, set_params, net_name='jones', tstop=170,
+                 dt=0.025, n_trials=1):
```
Comment (ntolley) [inline hnn_core/batch_simulate.py:16]: copy parameters and docstrings from `simulate_dipole` and pass options to `_run_single_sim()`

File: hnn_core/batch_simulate.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+            - `net`: The network model used for the simulation.
+            - `dpl`: The simulated dipole.
+            - `param_values`: The parameter values used for the simulation.
+        """
+        res = Parallel(n_jobs=n_jobs, verbose=50)(
+            delayed(self._run_single_sim)(
+                params) for params in param_combinations)
+        return res
+
+    def _run_single_sim(self, param_values):
```
Comment (ntolley) [inline hnn_core/batch_simulate.py:94]: In the next PR let's add an option to pass a callable in the `__init__()` that overrides this function

**Review by ntolley — COMMENTED** [review]
@samadpls I think this is almost ready if you want to switch the tag to [MRG]! Can you fix the flake8 errors and we'll move onto the next PR to work on outputting data files?

File: doc/whats_new.rst [hunk 17]
```diff
@@ -68,6 +68,9 @@ Changelog
 - Add function :func:`~hnn_core/params/convert_to_json` to convert legacy param
   and json files to new json format, by `George Dang`_ in :gh:`772`
 
+- Add :class:`~hnn_core.BatchSimulate` for batch simulation capability,
```
Comment (kmilo9999) [inline doc/whats_new.rst:71]: I think the right format is 
`~hnn_core/BatchSimulate`

File: .gitignore [hunk 18]
```diff
@@ -16,7 +16,8 @@ env/
 venv/
 ENV/
 VENV/
-
+hnn_core/mod/*
```
Comment (kmilo9999) [inline .gitignore:19]: I think this folder is already in the ignore list. Is there a reason to include it again?

File: .gitignore [hunk 19]
```diff
@@ -16,7 +16,8 @@ env/
...
+test.*
```
Comment (kmilo9999) [inline .gitignore:20]: I'm not sure about this. Why are you adding this type of files?

**Review by kmilo9999 — COMMENTED** [review]
I'm curious why the linkcheck is not passing. I did a few comments that might be helpful to solve it.

File: .gitignore (hunk 19 above)
Comment (samadpls) [reply .gitignore:20]: Apologies for the oversight; it was committed accidentally. Let me remove it

**Comment by ntolley** [comment]
@samadpls do you think this is related to the modifications we made to remove the GUI tab?

File: doc/whats_new.rst [hunk 20]
```diff
@@ -68,6 +68,9 @@ Changelog
 - Add function :func:`~hnn_core/params/convert_to_json` to convert legacy param
   and json files to new json format, by `George Dang`_ in :gh:`772`
 
+- Add :class:`~hnn_core/BatchSimulate` for batch simulation capability,
```
Comment (ntolley) [inline doc/whats_new.rst:71]:
```suggestion
- Add :class:`~hnn_core.BatchSimulate` for batch simulation capability,
```

Actually htis might be the solution!

**Comment by samadpls** [comment]
> @samadpls do you think this is related to the modifications we made to remove the GUI tab?

I'm not sure if it's related to the removal of the GUI tab since the other PR [(#754)](https://github.com/jonescompneurolab/hnn-core/pull/754) was handling that. In this PR, I haven't made any changes related to the GUI tab. I believe it might be because of the example file. I'm currently debugging it

File: hnn_core/batch_simulate.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
+"""Batch simulation."""
+
+# Authors: Abdul Samad Siddiqui <abdulsamadsid1@gmail.com>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+from joblib import Parallel, delayed
+from hnn_core import simulate_dipole
+from hnn_core.network_models import (jones_2009_model,
```
Comment (ntolley) [inline hnn_core/batch_simulate.py:10]:
```suggestion
from .network_models import (jones_2009_model,
```

File: hnn_core/batch_simulate.py [hunk 22]
```diff
@@ -0,0 +1,183 @@
+"""Batch simulation."""
+
+# Authors: Abdul Samad Siddiqui <abdulsamadsid1@gmail.com>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+from joblib import Parallel, delayed
+from hnn_core import simulate_dipole
```
Comment (ntolley) [inline hnn_core/batch_simulate.py:9]:
```suggestion
from .dipole import simulate_dipole
```

File: examples/howto/batch_simulate.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+for comprehensive analysis.
+"""
+
+# Authors: Abdul Samad Siddiqui <abdulsamadsid1@gmail.com>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+#############################################################################
+# Let us import ``hnn_core``.
```
Comment (ntolley) [inline examples/howto/batch_simulate.py:17]:
```suggestion
# Let us import hnn_core.
```

File: doc/api.rst [hunk 24]
```diff
@@ -94,6 +94,14 @@ Parallel backends (:py:mod:`hnn_core.parallel_backends`):
    MPIBackend
    JoblibBackend
 
+Batch Simulation (:py:mod:`hnn_core.batch_simulate`):
+---------------------------------------------------------
```
Comment (ntolley) [inline doc/api.rst:98]:
```suggestion
Batch Simulation (:py:mod:`hnn_core.batch_simulate`):
-----------------------------------------------------
```

**Comment by samadpls** [comment]
@ntolley, It seems like the linkcheck passes now. I removed the import statement
 ```python
from .batch_simulate import BatchSimulate
```
from the `__init__.py` file and the reference from `api.rst`.

**Comment by ntolley** [comment]
Really strange! There's definitely a step I'm forgetting in terms of adding a new file/class for importing.

In any case really great start @samadpls, going to go ahead and merge and we can continue to improve on this in the next PR

**Pull request merged**